File: lemur-app/src/main/java/io/github/malczuuu/lemur/app/infra/data/jpa/player/PlayerEntityMapper.java
package io.github.malczuuu.lemur.app.infra.data.jpa.player;

import io.github.malczuuu.lemur.app.common.IdAsLong;
import io.github.malczuuu.lemur.app.domain.DomainId;
import io.github.malczuuu.lemur.app.domain.player.Player;
import io.github.malczuuu.lemur.app.domain.player.PlayerBuilder;
import io.github.malczuuu.lemur.app.domain.player.PlayerStatus;

class PlayerEntityMapper {

  PlayerEntity toEntity(Player player) {
    PlayerEntity entity = new PlayerEntity();

    if (!player.isNew()) {
      IdAsLong id = IdAsLong.parse(player.getId().getValue());
      entity.setId(id.get());
    }

    entity.setName(player.getName());
    entity.setRating(player.getRating());
    entity.setStatus(player.getStatus().getLabel());
    player.getCreatedDate().ifPresent(entity::setCreatedDate);
    player.getLastModifiedDate().ifPresent(entity::setLastModifiedDate);
    entity.setVersion(player.getVersion());
    return entity;
  }

  Player toPlayer(PlayerEntity entity) {
    PlayerBuilder builder =
        new PlayerBuilder()
            .withId(DomainId.of(entity.getId()))
            .withName(entity.getName())
            .withRating(entity.getRating())
            .withStatus(PlayerStatus.parse(entity.getStatus()))
            .withCreatedDate(entity.getCreatedDate())
            .withLastModifiedDate(entity.getLastModifiedDate())
            .withVersion(entity.getVersion());
    return builder.build();
  }
}

File: lemur-app/src/main/kotlin/io/github/malczuuu/lemur/app/infra/data/jpa/player/PlayerEntityMapper.kt
package io.github.malczuuu.lemur.app.infra.data.jpa.player

import io.github.malczuuu.lemur.app.domain.DomainId
import io.github.malczuuu.lemur.app.domain.player.Player
import io.github.malczuuu.lemur.app.domain.player.PlayerStatus
import org.springframework.stereotype.Component
import kotlin.time.toJavaInstant
import kotlin.time.toKotlinInstant

@Component
internal class PlayerEntityMapper {

    fun toEntity(player: Player): PlayerEntity = PlayerEntity(
        _id = if (player.id.assigned) player.id.asLong() else null,
        name = player.name,
        rating = player.rating,
        status = player.status.label,
        createdDate = player.createdDate?.toJavaInstant(),
        lastModifiedDate = player.lastModifiedDate?.toJavaInstant(),
        version = player.version,
    )

    fun toPlayer(entity: PlayerEntity): Player = Player(
        id = DomainId(entity.id!!),
        name = entity.name,
        rating = entity.rating,
        status = PlayerStatus.parse(entity.status),
        createdDate = entity.createdDate?.toKotlinInstant(),
        lastModifiedDate = entity.lastModifiedDate?.toKotlinInstant(),
        version = entity.version!!,
    )
}
